src/Instruction.cpp

```diff
@@ -38,34 +42,47 @@ uint8_t Instruction::getZERO(uint32_t inst)
     return (inst) ? 0x0 : 0x1;
 }
 
-uint8_t Instruction::getCarry(uint64_t result)
+uint32_t Instruction::addWithCarry(uint32_t op1, uint32_t op2, uint32_t carryInAsBit)
 {
-    uint16_t highBit = (result >> 32) & 0xffffffff;
+    int carryIn = carryInAsBit ? 1 : 0;
+    uint64_t unsignedSum = (uint64_t)op1 + (uint64_t)op2 + (uint64_t)carryIn;
+    int64_t signedSum = (int64_t)(int32_t)op1 + (int64_t)(int32_t)op2 + (int64_t)carryIn;
+    uint32_t result = (uint32_t)unsignedSum;
 
-    if(highBit != 0x00)
-        return 1;
-    else
-        return 0;
-}
-uint8_t Instruction::getBorrow(uint64_t result)
-{
-    if(((result >> 32) & 0x01) == 0x00)
-        return 1;
-    else
-        return 0;
+    reg->PSR[psrC] = ((uint64_t)result == unsignedSum) ? 0 : 1;
+    reg->PSR[psrV] = ((int64_t)(int32_t)result == signedSum) ? 0 : 1;
+
+    return result;
 }
-uint8_t Instruction::getOverflow(uint32_t op1, uint32_t op2, uint32_t result)
+
+uint32_t Instruction::shiftC(uint32_t value, uint8_t type, uint32_t shift, uint32_t carryIn)
 {
-    if(getMSB(op1) == getMSB(op2))
+    uint32_t result = value;
+    reg->PSR[psrC] = carryIn;
+
+    if (shift != 0)
     {
-        if(getMSB(op1) != getMSB(result))
-            return 1;
-        else
-            return 0;
+        switch (type)
+        {
+        case LSL:
```

Comment:
This should be improved in https://github.com/whdlgp/ARMv6m_Simulator/pull/5/commits/e58d030599fc1f9d68e0ebe006034f5c6200f4dd

I uncommented more passing tests, ROR needs more work and I can see to it later on. For now this should address your concern. Thanks.